common/flatpak-dir.c

```diff
@@ -7548,6 +7567,112 @@ apply_extra_data (FlatpakDir   *self,
   return TRUE;
 }
 
+/* Check the user’s parental controls allow installation of @ref by looking at
+ * its cached @deploy_data, which contains its content rating as extracted from
+ * its AppData when it was originally downloaded. That’s compared to the
+ * parental controls policy loaded from the #MctManager.
+ *
+ * If @ref should not be installed, an error is returned. */
+static gboolean
+flatpak_dir_check_parental_controls (FlatpakDir    *self,
+                                     const char    *ref,
+                                     GVariant      *deploy_data,
+                                     GCancellable  *cancellable,
+                                     GError       **error)
+{
+#ifdef HAVE_LIBMALCONTENT
+  g_autoptr(GError) local_error = NULL;
+  const char *on_session = g_getenv ("FLATPAK_SYSTEM_HELPER_ON_SESSION");
+  g_autoptr(GDBusConnection) dbus_connection = NULL;
+  g_autoptr(MctManager) manager = NULL;
+  g_autoptr(MctAppFilter) app_filter = NULL;
+  const char *content_rating_type;
+  g_autoptr(GHashTable) content_rating = NULL;
+  g_autoptr(AutoPolkitAuthority) authority = NULL;
+  g_autoptr(AutoPolkitDetails) details = NULL;
+  g_autoptr(AutoPolkitSubject) subject = NULL;
+  g_autoptr(AutoPolkitAuthorizationResult) result = NULL;
+  gboolean authorized;
+
+  /* The ostree-metadata and appstream/ branches should not have any parental
+   * controls restrictions. */
+  if (!g_str_has_prefix (ref, "app/") && !g_str_has_prefix (ref, "runtime/"))
+    return TRUE;
+
+  g_debug ("Getting parental controls details for %s from %s",
+           ref, flatpak_deploy_data_get_origin (deploy_data));
+
+  dbus_connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, cancellable, &local_error);
+  if (dbus_connection == NULL)
+    {
+      if (on_session != NULL)
+        {
+          /* FIXME: Instead of skipping the parental controls check in the test
+           * environment, make a mock service for it.
+           * https://phabricator.endlessm.com/T25340 */
+          g_debug ("Skipping parental controls check for %s since the "
+                   "system bus is unavailable in the test environment", ref);
+          return TRUE;
+        }
+      else
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+    }
+
+  manager = mct_manager_new (dbus_connection);
+  app_filter = mct_manager_get_app_filter (manager, getuid (),
+                                           MCT_GET_APP_FILTER_FLAGS_INTERACTIVE,
+                                           cancellable, error);
+  if (app_filter == NULL)
+    return FALSE;
+
```

Comment:
I feel that we should maybe handle a failure here differently. If we are linking to libmalcontent, and we have a system bus connection, which we (somewhat) validated as trustworthy, and it turns out that there is no malcontent manager running on the system then I think this should be equivalent with "everything is allowed".

This means is should be possible to ship a flatpak linked to libmalcontent, but then chose not to install the manager and still have everything work the same as if libmalcontent was not linked to.

Does this make sense?

Reply:
Ah, this is a good catch. I actually added support for that in libmalcontent 0.3.0, but forgot to update the flatpak branch to use it. malcontent can return `MCT_APP_FILTER_ERROR_DISABLED` to indicate that filtering is disabled system-wide because, for example, the manager isn’t installed.

Fixed locally.